x/qgb/orchestrator/parser.go

```diff
@@ -0,0 +1,96 @@
+package orchestrator
+
+import (
+	"fmt"
+
+	"github.com/celestiaorg/celestia-app/app"
+	qgbtypes "github.com/celestiaorg/celestia-app/x/qgb/types"
+	sdkcodec "github.com/cosmos/cosmos-sdk/codec"
+	sdkcodectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/std"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/gogo/protobuf/proto"
+	coretypes "github.com/tendermint/tendermint/types"
+)
+
+var _ QGBParserI = QGBParser{}
+
+type QGBParserI interface {
+	IsDataCommitmentConfirm(msg sdktypes.Msg) (bool, error)
+	ParseDataCommitmentConfirm(msg sdktypes.Msg) (qgbtypes.MsgDataCommitmentConfirm, error)
+	IsValsetConfirm(msg sdktypes.Msg) (bool, error)
+	ParseValsetConfirm(msg sdktypes.Msg) (qgbtypes.MsgValsetConfirm, error)
+	ParseCoreTx(tx coretypes.Tx) (sdktx.Tx, error)
+	ParseSdkTx(any *sdkcodectypes.Any) (sdktypes.Msg, error)
+}
+
+type QGBParser struct {
+	codec sdkcodec.Codec
+}
+
+func NewQGBParser(codec sdkcodec.Codec) *QGBParser {
+	return &QGBParser{codec: codec}
+}
+
+func (parser QGBParser) IsValsetConfirm(msg sdktypes.Msg) (bool, error) {
+	switch msg.(type) {
+	case *qgbtypes.MsgValsetConfirm:
+		return true, nil
+	default:
+		return false, nil
+	}
+}
+
+func (parser QGBParser) IsDataCommitmentConfirm(msg sdktypes.Msg) (bool, error) {
+	switch msg.(type) {
+	case *qgbtypes.MsgDataCommitmentConfirm:
+		return true, nil
+	default:
+		return false, nil
+	}
+}
+
+func (parser QGBParser) ParseValsetConfirm(msg sdktypes.Msg) (qgbtypes.MsgValsetConfirm, error) {
+	vs, ok := msg.(*qgbtypes.MsgValsetConfirm)
+	if !ok {
+		return qgbtypes.MsgValsetConfirm{},
+			fmt.Errorf("unable to cast sdk message to msg valset confirm")
+	}
+	return *vs, nil
+}
+
+func (parser QGBParser) ParseDataCommitmentConfirm(msg sdktypes.Msg) (qgbtypes.MsgDataCommitmentConfirm, error) {
+	dcc, ok := msg.(*qgbtypes.MsgDataCommitmentConfirm)
+	if !ok {
+		return qgbtypes.MsgDataCommitmentConfirm{},
+			fmt.Errorf("unable to cast sdk message to msg data commitment confirm")
+	}
+	return *dcc, nil
+}
+
+func (parser QGBParser) ParseCoreTx(tx coretypes.Tx) (sdktx.Tx, error) {
+	var sdkTx sdktx.Tx
+	err := proto.Unmarshal(tx, &sdkTx)
+	if err != nil {
+		return sdktx.Tx{}, fmt.Errorf("error while unmarshalling core transaction: %s", err)
+	}
+	return sdkTx, nil
+}
+
+func (parser QGBParser) ParseSdkTx(any *sdkcodectypes.Any) (sdktypes.Msg, error) {
+	var stdMsg sdktypes.Msg
+	err := parser.codec.UnpackAny(any, &stdMsg)
+	if err != nil {
+		return nil, fmt.Errorf("error while unpacking message: %s", err)
+	}
+	return stdMsg, nil
+}
+
+func MakeDefaultAppCodec() sdkcodec.Codec {
+	interfaceRegistry := sdkcodectypes.NewInterfaceRegistry()
+	std.RegisterInterfaces(interfaceRegistry)
+	app.ModuleBasics.RegisterInterfaces(interfaceRegistry)
+	qgbtypes.RegisterInterfaces(interfaceRegistry)
+	return sdkcodec.NewProtoCodec(interfaceRegistry)
+}
```

Comment:
the only time we don't is during testing for packages that are imported by the app package (or that is the only time that we should be ideally, like in the [payment](https://github.com/celestiaorg/celestia-app/blob/c9dce391bc360bb5f068f43d4eb1681a9f35b710/x/payment/types/codec.go#L51-L57) module). If this one is just for testing and we need the orchestrator package imported by the app package, then we should unexport this one.

Reply:
This one is not for testing. It is used to decode messages that the ingestion receives.
Is there a place where I can get the app encoding without creating it myself here?
If so, that would be even better.

Reply:
We create the encoding config each time to avoid requiring a global or access to the app instance itself, so unfortunately we just have to create it using the link in the comment above

Reply:
So, I create something like this on the state machine side:
```go
type localEncoder struct{}

func (localEncoder) RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
	RegisterCodec(cdc)
}

func (localEncoder) RegisterInterfaces(r codectypes.InterfaceRegistry) {
	RegisterInterfaces(r)
}

func MakeQGBEncodingConfig() encoding.Config {
	e := localEncoder{}
	return encoding.MakeConfig(e)
}
```
and use it on the ingestion side?

Reply:
8652ce9